### PR TITLE
support object fields in aggregation based sigma rules

### PR DESCRIPTION
**Pull request by sbcd90**
### Description
support object fields in aggregation based sigma rules.
add support for `count(field)` in aggregation based sigma rules, count(*) is currently supported.
e.g.
```
id: 25b9c01c-350d-4b95-bed1-836d04a4f324
title: test
description: Detects when an user creates or invokes a lambda function.
status: experimental
author: test
date: 2023/12/07
modified: 2023/12/07
logsource:
  category: cloudtrail
level: low
detection:
  condition: selection1 or selection2 | count(api.operation) by cloud.region > 1
  selection1:
    api.service.name:
      - lambda.amazonaws.com
    api.operation:
      - CreateFunction
  selection2:
    api.service.name:
      - lambda.amazonaws.com
    api.operation:      
      - Invoke
  timeframe: 20m
  tags:
    - attack.privilege_escalation
    - attack.t1078
```
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/grammars/Aggregation.g4 [hunk 1]
```diff
@@ -21,7 +21,7 @@ RPAREN : ')' ;
 
 DECIMAL : '-'?[0-9]+('.'[0-9]+)? ;
 
-IDENTIFIER : [a-zA-Z*_][a-zA-Z_0-9]* ;
+IDENTIFIER : [a-zA-Z*_.][a-zA-Z_0-9.]* ;
```
Comment (eirsep) [inline src/main/grammars/Aggregation.g4:24]: will this work for `nested` type fields also?

File: src/main/generated/org/opensearch/securityanalytics/rules/condition/aggregation/AggregationLexer.java [hunk 2]
```diff
@@ -129,50 +129,50 @@ public AggregationLexer(CharStream input) {
...
-        "\u0000\u0000WX\u0001\u0000\u0000\u0000XZ\u0001\u0000\u0000\u0000YS\u0001"+
-        "\u0000\u0000\u0000YZ\u0001\u0000\u0000\u0000Z\u001c\u0001\u0000\u0000"+
-        "\u0000[_\u0007\u0001\u0000\u0000\\^\u0007\u0002\u0000\u0000]\\\u0001\u0000"+
-        "\u0000\u0000^a\u0001\u0000\u0000\u0000_]\u0001\u0000\u0000\u0000_`\u0001"+
-        "\u0000\u0000\u0000`\u001e\u0001\u0000\u0000\u0000a_\u0001\u0000\u0000"+
-        "\u0000bd\u0007\u0003\u0000\u0000cb\u0001\u0000\u0000\u0000de\u0001\u0000"+
-        "\u0000\u0000ec\u0001\u0000\u0000\u0000ef\u0001\u0000\u0000\u0000fg\u0001"+
-        "\u0000\u0000\u0000gh\u0006\u000f\u0000\u0000h \u0001\u0000\u0000\u0000"+
-        "\u0007\u0000LQWY_e\u0001\u0006\u0000\u0000";
+        "\u0000\u0004\u0001\u000009\u0005\u0000**..AZ__az\u0005\u0000..09AZ__a"+
```
Comment (eirsep) [inline src/main/generated/org/opensearch/securityanalytics/rules/condition/aggregation/AggregationLexer.java:132]: what are these changes?

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java [hunk 3]
```diff
@@ -376,17 +376,17 @@ public AggregationQueries convertAggregation(AggregationItem aggregation) {
             Script script = new Script(String.format(Locale.getDefault(), bucketTriggerScript, "_cnt", aggregation.getCompOperator(), aggregation.getThreshold()));
             condition = new BucketSelectorExtAggregationBuilder(bucketTriggerSelectorId, Collections.singletonMap("_cnt", "_count"), script, "result_agg", null);
         } else {
-            fmtAggQuery = String.format(Locale.getDefault(), aggQuery, "result_agg", aggregation.getGroupByField(), aggregation.getAggField(), aggregation.getAggFunction(), aggregation.getAggField());
-            fmtBucketTriggerQuery = String.format(Locale.getDefault(), bucketTriggerQuery, aggregation.getAggField(), aggregation.getAggField(), "result_agg", aggregation.getAggField(), aggregation.getCompOperator(), aggregation.getThreshold());
+            fmtAggQuery = String.format(Locale.getDefault(), aggQuery, "result_agg", aggregation.getGroupByField(), aggregation.getAggField().replace(".", "_"), aggregation.getAggFunction().equals("count")? "value_count": aggregation.getAggFunction(), aggregation.getAggField());
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java:379]: why have we changed from `aggregation.getAggFunction()` to `aggregation.getAggFunction().equals("count")? "value_count": aggregation.getAggFunction()`?
Is this intentional?

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java [hunk 4]
```diff
@@ -376,17 +376,17 @@ public AggregationQueries convertAggregation(AggregationItem aggregation) {
...
+            fmtBucketTriggerQuery = String.format(Locale.getDefault(), bucketTriggerQuery, aggregation.getAggField().replace(".", "_"), aggregation.getAggField(), "result_agg", aggregation.getAggField().replace(".", "_"), aggregation.getCompOperator(), aggregation.getThreshold());
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java:380]: plz do `aggregation.getAggField().replace(".", "_")` **only once** and store in variable and re-use everywhere else

File: src/test/java/org/opensearch/securityanalytics/TestHelpers.java [hunk 5]
```diff
@@ -1789,6 +1833,106 @@ public static String randomAdLdapDoc() {
                 "}";
     }
 
+    public static String randomCloudtrailOcsfDoc() {
```
Comment (eirsep) [inline src/test/java/org/opensearch/securityanalytics/TestHelpers.java:1836]: can you create a document with object type fields in `a.b.c` (dot) format and test the same agg rule?

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java [hunk 6]
```diff
@@ -361,9 +361,9 @@ public AggregationQueries convertAggregation(AggregationItem aggregation) {
         BucketSelectorExtAggregationBuilder condition;
         String bucketTriggerSelectorId = UUIDs.base64UUID();
 
-        if (aggregation.getAggFunction().equals("count")) {
+        if (aggregation.getAggFunction().equals("count") && aggregation.getAggField().equals("*")) {
```
Comment (eirsep) [inline src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java:364]: is there any edge case where field mappings need to be considered during aggregation rule parsing?

what if customer creates an aggregation rule with ECS field name?

can you add a test case where customer use ecs field name in rule and automatically mapped raw field name in index mapping

**Review by eirsep — CHANGES_REQUESTED** [review]

File: src/main/grammars/Aggregation.g4 (hunk 1 above)
Comment (sbcd90) [reply src/main/grammars/Aggregation.g4:24]: `nested` fields cannot be queried using `query string queries`.

File: src/main/generated/org/opensearch/securityanalytics/rules/condition/aggregation/AggregationLexer.java (hunk 2 above)
Comment (sbcd90) [reply src/main/generated/org/opensearch/securityanalytics/rules/condition/aggregation/AggregationLexer.java:132]: these are generated code by antlr.

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java (hunk 3 above)
Comment (sbcd90) [reply src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java]: we want to support `count(field)`. earlier, we only supported `count(*)`.

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java (hunk 4 above)
Comment (sbcd90) [reply src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java]: fixed it.

File: src/test/java/org/opensearch/securityanalytics/TestHelpers.java (hunk 5 above)
Comment (sbcd90) [reply src/test/java/org/opensearch/securityanalytics/TestHelpers.java:1872]: ok updated the mapping to an object field. https://github.com/opensearch-project/security-analytics/pull/789/files#diff-32ce4f632f4dca5253d801824d6ac90e9fa458a73bf21a2a04f57750f98fd5c4R983

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java (hunk 6 above)
Comment (sbcd90) [reply src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java:364]: added an aggregation rule which leverages `raw->ecs mapping`.

File: src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java (hunk 3 above)
Comment (eirsep) [reply src/main/java/org/opensearch/securityanalytics/rules/backend/OSQueryBackend.java]: plz update description of PR to call this out.

**Review by eirsep — APPROVED** [review]
